### ISSUE TITLE
Feature request: expose FFI API

### ISSUE THREAD

**Issue by etra0** [veeenu/hudhook#58]
I was looking for IMGUI hooks in general, and found kiero, which is more to hook to DirectX APIs, and a few libraries that use kiero to hook IMGUI, but none of them were as well capsulated as this one and one-for-all apis.

Since I like experimenting with languages (I still enjoy Rust tho!) I was thinking it may be possible to expose the IMGUI hook to FFI to be able to use it with other languages.

I think we only should need to expose very few functions (per Implementation) and it would be not so complicated.

I think the most troublesome could be expose imgui itself, but I wouldn't say it's impossible.

Thoughts?

**Comment by veeenu** [comment]
Ironically, the state of the available overlay library projects was the reason that led me to developing hudhook in the first place 😅 I know Rust is a lot more opinionated than other languages and doesn't make it easy to interop with other languages, but that's a conscious tradeoff I've made when I moved to it from C++.

Making bindings to hudhook should be relatively straightforward, just `#[no_mangle] extern "C"` functions, though due to the "injection" nature we should find a way to still initialize it in some `DllMain`. I'm honestly not sure of the implications of exposing Rust code to be used as FFI library in terms of compiler assumptions. Doing it in already established formats (see e.g. pyo3 or wasm) comes with a number of caveats, no idea what would happen with code like hudhook whose bread and butter is skirting right along the line of undefined behavior.

You pointed out a valid concern: I think it would be super hard to expose the imgui API, for example due to its reliance on global statics; say, compiling another copy of imgui would result in a different set of global statics and I have no idea what that would entail or how to have those play nice together. In the absence of that, exposing C wrappers to imgui-rs's api which is already a wrapper over cimgui which is itself a wrapper around imgui would probably be a total maintenance nightmare.

If anything, I'd be ok with an external crate that depends on hudhook and reexports its API, but I don't think I'd be comfortable in maintaining that upstream. It would add another dimension to maintenance and would probably be a bigger project than hudhook itself.
